File: 3659-count-paths-with-the-given-xor-value/3659-count-paths-with-the-given-xor-value.java
class Solution {
    final int MOD = (int) 1e9 + 7;

    private int help(int[][] grid, int n, int m, int val, int k, int[][][] dp) {
        if (n == 0 && m == 0) {
            if ((val ^ grid[0][0]) == k)
                return 1;
            else
                return 0;
        }
        if (n < 0 || m < 0)
            return 0;
        if (dp[n][m][val] != -1)
            return dp[n][m][val];
        int up = help(grid, n - 1, m, val ^ grid[n][m], k, dp);
        int left = help(grid, n, m - 1, val ^ grid[n][m], k, dp);

        return dp[n][m][val] = (up + left) % MOD;

    }

    public int countPathsWithXorValue(int[][] grid, int k) {
        int n = grid.length;
        int m = grid[0].length;
        int[][][] dp = new int[n][m][16];
        for (int r = 0; r < n; r++)
            for (int s = 0; s < m; s++)
                for (int t = 0; t < 16; t++)
                    dp[r][s][t] = -1;

        return help(grid, n - 1, m - 1, 0, k, dp);
    }
}

File: 3659-count-paths-with-the-given-xor-value/3659-count-paths-with-the-given-xor-value.cs
public class Solution {
    const int MOD = 1000000007;
    const int MAX_XOR_VALUE = 16;

    public int Solve(int[][] grid, int i, int j, int val, int k, List<List<List<int>>> dp) {
        if (i == 0 && j == 0)
            return (val ^ grid[0][0]) == k ? 1 : 0;

        if (i < 0 || j < 0)
            return 0;

        if (dp[i][j][val] != -1)
            return dp[i][j][val];


        int up = Solve(grid, i - 1, j, val ^ grid[i][j], k, dp);  
        int left = Solve(grid, i, j - 1, val ^ grid[i][j], k, dp);  

        return dp[i][j][val] = (up + left) % MOD;
    }

    public int CountPathsWithXorValue(int[][] grid, int k) {
        int n = grid.Length;
        int m = grid[0].Length;

        List<List<List<int>>> dp = new List<List<List<int>>>();
        for (int i = 0; i < n; i++) {
            List<List<int>> row = new List<List<int>>();
            for (int j = 0; j < m; j++) {
                List<int> valList = new List<int>(new int[MAX_XOR_VALUE]);
                for (int l = 0; l < MAX_XOR_VALUE; l++) {
                    valList[l] = -1;
                }
                row.Add(valList);
            }
            dp.Add(row);
        }

        int ans = Solve(grid, n - 1, m - 1, 0, k, dp);

        return ans % MOD;
    }
}